File: Cloud Keys Samples/Google Secrets Manager/java/googlesecrets.java
/*
 * Cloud Keys 2024 Java Edition - Sample Project
 *
 * This sample project demonstrates the usage of Cloud Keys in a 
 * simple, straightforward way. It is not intended to be a complete 
 * application. Error handling and other checks are simplified for clarity.
 *
 * www.nsoftware.com/cloudkeys
 *
 * This code is subject to the terms and conditions specified in the 
 * corresponding product license agreement which outlines the authorized 
 * usage and restrictions.
 */

import java.io.*;
import cloudkeys.*;

public class googlesecrets extends ConsoleDemo{

	static String command; // user's command
	static String buffer; // text buffer
	static String[] argument; // arguments following command

	static String secretName;
	static String secretType;
	static String data;

	public static void main(String[] args) {

		GoogleSecrets googlesecrets = new GoogleSecrets();
		try {
			googlesecrets.addGoogleSecretsEventListener(new DefaultGoogleSecretsEventListener(){

				@Override
				public void error(GoogleSecretsErrorEvent e) {
					System.out.println("Error [" + e.errorCode + "]: " + e.description);
				}

				@Override
				public void secretList(GoogleSecretsSecretListEvent e) {
					System.out.println("   " + e.name);
				}
			});

			System.out.println("Press Enter to Authenticate.");
			input();

			googlesecrets.getOAuth().setClientId("157623334268.apps.googleusercontent.com");
			googlesecrets.getOAuth().setClientSecret("v8R8R90hn_LchsVc0Ta6Sy0D");
			googlesecrets.getOAuth().setAuthorizationScope("https://www.googleapis.com/auth/cloud-platform");
			googlesecrets.getOAuth().setServerAuthURL("https://accounts.google.com/o/oauth2/auth");
			googlesecrets.getOAuth().setServerTokenURL("https://accounts.google.com/o/oauth2/token");

			googlesecrets.authorize();
			System.out.println("Authentication Successful.");

			googlesecrets.setGoogleProjectId(prompt("Google Cloud Project ID"));

			while (true) {
				DisplayMenu();
				System.out.print("> ");
				command = input();
				argument = command.split("\\s");
				if (argument.length == 0)
					continue;
				switch (argument[0].charAt(0)) {
					case 'l':
						System.out.println("Secrets: ");
						googlesecrets.listSecrets();
						System.out.println();
						break;
					case 'c':
						secretName = prompt("Secret Name", ":", "TestSecret");
						data = prompt("Secret Data", ":", "Test");

						googlesecrets.setSecretData(data);
						googlesecrets.createSecret(secretName);

						System.out.println("Secret Created Successfully");
						break;
					case 'd':
						secretName = prompt("Secret Name", ":", "TestSecret");
						googlesecrets.deleteSecret(secretName);
						System.out.println("Secret Deleted Successfully");
						break;
					case 'v':
						secretName = prompt("Secret Name", ":", "TestSecret");
						googlesecrets.getSecret(secretName, "");

						System.out.println("Secret Data: " + new String(googlesecrets.getSecretData()));
						break;
					case '?':
						DisplayMenu();
						break;
					case 'q':
						return;
					default:
						System.out.println("Command Not Recognized.");
				}
			}

		} catch (Exception ex) {
			System.out.println(ex.getMessage());
		}
	}

	private static void DisplayMenu() {
			System.out.println("\nGoogleSecrets Commands:");
			System.out.println("l            List Secrets");
			System.out.println("c            Create a Secret");
			System.out.println("d            Delete a Secret");
			System.out.println("v            View a Secret's Data");
			System.out.println("?            Display Options");
			System.out.println("q            Quit");
	}
}

class ConsoleDemo {
  private static BufferedReader bf = new BufferedReader(new InputStreamReader(System.in));

  static String input() {
    try {
      return bf.readLine();
    } catch (IOException ioe) {
      return "";
    }
  }
  static char read() {
    return input().charAt(0);
  }

  static String prompt(String label) {
    return prompt(label, ":");
  }
  static String prompt(String label, String punctuation) {
    System.out.print(label + punctuation + " ");
    return input();
  }
  static String prompt(String label, String punctuation, String defaultVal) {
      System.out.print(label + " [" + defaultVal + "]" + punctuation + " ");
      String response = input();
      if (response.equals(""))
        return defaultVal;
      else
        return response;
  }

  static char ask(String label) {
    return ask(label, "?");
  }
  static char ask(String label, String punctuation) {
    return ask(label, punctuation, "(y/n)");
  }
  static char ask(String label, String punctuation, String answers) {
    System.out.print(label + punctuation + " " + answers + " ");
    return Character.toLowerCase(read());
  }

  static void displayError(Exception e) {
    System.out.print("Error");
    if (e instanceof CloudKeysException) {
      System.out.print(" (" + ((CloudKeysException) e).getCode() + ")");
    }
    System.out.println(": " + e.getMessage());
    e.printStackTrace();
  }

  /**
   * Takes a list of switch arguments or name-value arguments and turns it into a map.
   */
  static java.util.Map<String, String> parseArgs(String[] args) {
    java.util.Map<String, String> map = new java.util.HashMap<String, String>();
    
    for (int i = 0; i < args.length; i++) {
      // Add a key to the map for each argument.
      if (args[i].startsWith("-")) {
        // If the next argument does NOT start with a "-" then it is a value.
        if (i + 1 < args.length && !args[i + 1].startsWith("-")) {
          // Save the value and skip the next entry in the list of arguments.
          map.put(args[i].toLowerCase().replaceFirst("^-+", ""), args[i + 1]);
          i++;
        } else {
          // If the next argument starts with a "-", then we assume the current one is a switch.
          map.put(args[i].toLowerCase().replaceFirst("^-+", ""), "");
        }
      } else {
        // If the argument does not start with a "-", store the argument based on the index.
        map.put(Integer.toString(i), args[i].toLowerCase());
      }
    }
    return map;
  }
}

File: Cloud Keys Samples/Google Secrets Manager/net/googlesecrets.cs
/*
 * Cloud Keys 2024 .NET Edition - Sample Project
 *
 * This sample project demonstrates the usage of Cloud Keys in a 
 * simple, straightforward way. It is not intended to be a complete 
 * application. Error handling and other checks are simplified for clarity.
 *
 * www.nsoftware.com/cloudkeys
 *
 * This code is subject to the terms and conditions specified in the 
 * corresponding product license agreement which outlines the authorized 
 * usage and restrictions.
 * 
 */

﻿using System;
using nsoftware.CloudKeys;

class googlesecretsDemo
{
  private static GoogleSecrets googlesecrets = new nsoftware.CloudKeys.GoogleSecrets();

  static void Main(string[] args)
  {
    try
    {
      googlesecrets.OnSecretList += googlesecrets_OnSecretList;
      googlesecrets.OnError += googlesecrets_OnError;

      // Prompt for authentication information.
      Console.Write("Enter your OAuth client ID: "); // Can use 157623334268.apps.googleusercontent.com for testing purposes.
      googlesecrets.OAuth.ClientId = Console.ReadLine();
      Console.Write("Enter your OAuth client secret: "); // Can use v8R8R90hn_LchsVc0Ta6Sy0D for testing purposes.
      googlesecrets.OAuth.ClientSecret = Console.ReadLine();

      googlesecrets.OAuth.ServerAuthURL = "https://accounts.google.com/o/oauth2/auth";
      googlesecrets.OAuth.ServerTokenURL = "https://accounts.google.com/o/oauth2/token";
      googlesecrets.OAuth.AuthorizationScope = "https://www.googleapis.com/auth/cloud-platform";

      googlesecrets.Config("OAuthBrowserResponseTimeout=60");
      googlesecrets.Authorize();

      Console.Write("Enter your Google Cloud project ID: ");
      googlesecrets.GoogleProjectId = Console.ReadLine();

      // Process user commands.
      Console.WriteLine("Type \"?\" or \"help\" for a list of commands.");
      Console.Write("googlesecrets> ");
      string command;
      string[] arguments;

      while (true)
      {
        command = Console.ReadLine();
        arguments = command.Split();

        if (arguments[0] == "?" || arguments[0] == "help")
        {
          Console.WriteLine("Commands: ");
          Console.WriteLine("  ?                               display the list of valid commands");
          Console.WriteLine("  help                            display the list of valid commands");
          Console.WriteLine("  ls                              list secrets");
          Console.WriteLine("  create <name>                   create a new secret with the specified name");
          Console.WriteLine("  del <name>                      delete the secret with the specified name");
          Console.WriteLine("  view <name>                     view a secret's data");
          Console.WriteLine("  quit                            exit the application");
        }
        else if (arguments[0] == "ls")
        {
          googlesecrets.ListSecrets();
        }
        else if (arguments[0] == "create")
        {
          if (arguments.Length > 1)
          {
            Console.Write("Enter the secret data: ");
            googlesecrets.SecretData = Console.ReadLine();

            googlesecrets.CreateSecret(arguments[1]);
            Console.WriteLine("Secret created successfully.");
          }
        }
        else if (arguments[0] == "del")
        {
          if (arguments.Length > 1)
          {
            googlesecrets.DeleteSecret(arguments[1]);
            Console.WriteLine("Secret deleted successfully.");
          }
        }
        else if (arguments[0] == "view")
        {
          if (arguments.Length > 1)
          {
            googlesecrets.GetSecret(arguments[1], "");
            Console.WriteLine("Secret data: " + googlesecrets.SecretData);
          }
        }
        else if (arguments[0] == "quit")
        {
          break;
        }
        else if (arguments[0] == "")
        {
          // Do nothing.
        }
        else
        {
          Console.WriteLine("Invalid command.");
        } // End of command checking.

        Console.Write("googlesecrets> ");
      }
    }
    catch (Exception ex)
    {
      Console.WriteLine(ex.Message);
    }
  }

  private static void googlesecrets_OnSecretList(object sender, GoogleSecretsSecretListEventArgs e)
  {
    Console.WriteLine("   " + e.Name);
  }

  private static void googlesecrets_OnError(object sender, GoogleSecretsErrorEventArgs e)
  {
    Console.WriteLine("Error [" + e.ErrorCode + "]: " + e.Description);
  }
}




class ConsoleDemo
{
  /// <summary>
  /// Takes a list of switch arguments or name-value arguments and turns it into a dictionary.
  /// </summary>
  public static System.Collections.Generic.Dictionary<string, string> ParseArgs(string[] args)
  {
    System.Collections.Generic.Dictionary<string, string> dict = new System.Collections.Generic.Dictionary<string, string>();

    for (int i = 0; i < args.Length; i++)
    {
      // Add a key to the dictionary for each argument.
      if (args[i].StartsWith("/"))
      {
        // If the next argument does NOT start with a "/", then it is a value.
        if (i + 1 < args.Length && !args[i + 1].StartsWith("/"))
        {
          // Save the value and skip the next entry in the list of arguments.
          dict.Add(args[i].ToLower().TrimStart('/'), args[i + 1]);
          i++;
        }
        else
        {
          // If the next argument starts with a "/", then we assume the current one is a switch.
          dict.Add(args[i].ToLower().TrimStart('/'), "");
        }
      }
      else
      {
        // If the argument does not start with a "/", store the argument based on the index.
        dict.Add(i.ToString(), args[i].ToLower());
      }
    }
    return dict;
  }
  /// <summary>
  /// Asks for user input interactively and returns the string response.
  /// </summary>
  public static string Prompt(string prompt, string defaultVal)
  {
    Console.Write(prompt + (defaultVal.Length > 0 ? " [" + defaultVal + "]": "") + ": ");
    string val = Console.ReadLine();
    if (val.Length == 0) val = defaultVal;
    return val;
  }
}